From a5eab92289b2f800597be788e255c0827695e49c Mon Sep 17 00:00:00 2001
From: Robin Gong <yibin.gong@nxp.com>
Date: Wed, 22 Feb 2017 16:52:51 +0800
Subject: [PATCH 1541/5242] MLK-14241-3 ARM: imx: pm-rpmsg: cancel heartbeat
 workqueue before suspend

commit  3a17aa94711803836c556751e61bfd0772c2fe05 from
https://source.codeaurora.org/external/imx/linux-imx.git

cancel heartbeat workqueue to make sure no any heartbeat message will
be sent to M4 after M4 thought A7 core enter VLLS mode.

Signed-off-by: Robin Gong <yibin.gong@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 arch/arm/mach-imx/pm-rpmsg.c |   21 +++++++++++++++++++--
 1 file changed, 19 insertions(+), 2 deletions(-)

diff --git a/arch/arm/mach-imx/pm-rpmsg.c b/arch/arm/mach-imx/pm-rpmsg.c
index 4cd9262..450687a 100644
--- a/arch/arm/mach-imx/pm-rpmsg.c
+++ b/arch/arm/mach-imx/pm-rpmsg.c
@@ -248,12 +248,29 @@ static void pm_rpmsg_remove(struct rpmsg_device *rpdev)
 #ifdef CONFIG_PM_SLEEP
 static int pm_heartbeat_suspend(struct device *dev)
 {
-	return pm_vlls_notify_m4(true);
+	int err;
+
+	err = pm_vlls_notify_m4(true);
+	if (err)
+		return err;
+
+	cancel_delayed_work_sync(&heart_beat_work);
+
+	return 0;
 }
 
 static int pm_heartbeat_resume(struct device *dev)
 {
-	return pm_vlls_notify_m4(false);
+	int err;
+
+	err = pm_vlls_notify_m4(true);
+	if (err)
+		return err;
+
+	schedule_delayed_work(&heart_beat_work,
+			msecs_to_jiffies(10000));
+
+	return 0;
 }
 #endif
 
-- 
1.7.9.5

